source/particle/integrator/rk4_38.cc

```diff
@@ -0,0 +1,287 @@
+        property_indices[3] = property_indices[2] + dim;
+      }
+
+
+
+      template <int dim>
+      void
+      RK4_38<dim>::local_integrate_step(const typename ParticleHandler<dim>::particle_iterator &begin_particle,
+                                     const typename ParticleHandler<dim>::particle_iterator &end_particle,
+                                     const std::vector<Tensor<1,dim>> &old_velocities,
+                                     const std::vector<Tensor<1,dim>> &velocities,
+                                     const double dt)
+      {
+        if (integrator_substep < 3)
+          {
+            Assert(static_cast<unsigned int> (std::distance(begin_particle, end_particle)) == old_velocities.size(),
+                   ExcMessage("The particle integrator expects the old velocity vector to be of equal size "
+                              "to the number of particles to advect. For some unknown reason they are different, "
+                              "most likely something went wrong in the calling function."));
+          }
+
+        if (integrator_substep >= 1 && integrator_substep < 4)
+          {
+            Assert(static_cast<unsigned int> (std::distance(begin_particle, end_particle)) == velocities.size(),
+                   ExcMessage("The particle integrator expects the velocity vector to be of equal size "
+                              "to the number of particles to advect. For some unknown reason they are different, "
+                              "most likely something went wrong in the calling function."));
+          }
+
+        const auto cell = begin_particle->get_surrounding_cell();
+        bool at_periodic_boundary = false;
+        if (this->get_triangulation().get_periodic_face_map().empty() == false)
+          for (const auto &face_index: cell->face_indices())
+            if (cell->at_boundary(face_index))
+              if (cell->has_periodic_neighbor(face_index))
+                {
+                  at_periodic_boundary = true;
+                  break;
+                }
+
+        typename std::vector<Tensor<1,dim>>::const_iterator old_velocity = old_velocities.begin();
+        typename std::vector<Tensor<1,dim>>::const_iterator velocity = velocities.begin();
+
+        std::array<Tensor<1,dim>,4> k;
+        for (typename ParticleHandler<dim>::particle_iterator it = begin_particle;
+             it != end_particle; ++it, ++velocity, ++old_velocity)
+          {
+            ArrayView<double> properties = it->get_properties();
+
+            if (integrator_substep == 0)
+              {
+#if DEAL_II_VERSION_GTE(9, 6, 0)
+                // Get a reference to the particle location, so that we can update it in-place
+                Point<dim> &location = it->get_location();
+#else
+                Point<dim> location = it->get_location();
+#endif
+                k[0] = dt * (*old_velocity);
+
+                Point<dim> new_location = location + 1./3. * k[0];
```

Comment:
Here is a little trick that can become important, because particle advection happens so often: It is much faster for a CPU to do multiplication than division. Because you use `1./3.` and `2./3.` so often and the result is always the same, it can be useful to exclude it from the loop over all particles and compute it once in line 88:
```
const double one_thirds = 1./3.;
const double two_thirds = 2./3.;
```
Then you can do here and elsewhere:
```suggestion
                Point<dim> new_location = location + one_thirds * k[0];
```